template/registry/registry.go

```diff
@@ -0,0 +1,146 @@
+package registry
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+type Registry map[string]Func
+
+// Func is a function that can be used in formatting and matching contexts.
+type Func struct {
+	FormatSignature *Signature                             // Signature of the function when called in formatting context
+	MatchSignature  *Signature                             // Signature of the function when called in matching context
+	Fn              func(any, map[string]any) (any, error) // Function itself
+	Name            string
+	Description     string
+}
+
+// Signature is a signature of the function, i.e. what input and options are allowed.
+type Signature struct {
+	ValidateInput   func(any) error
+	Options         Options
+	IsInputRequired bool
+}
+
+// Option is a possible options for the function.
+type Option struct {
+	Name        string
+	Description string
+	Default     any
+
+	ValidateValue  func(any) error // If option value is not restricted by a set of values.
+	PossibleValues []any           // If option value is restricted by a set of values.
+}
+
+type Options []Option
+
+// NewRegistry returns a new registry with default functions.
+func NewRegistry() Registry {
```

Comment:
Consumers use `registry.New()`. This package unlikely to have other `New` functions.

```suggestion
func New() Registry {
```